keps/sig-api-machinery/20190426-crd-defaulting.md

```diff
@@ -0,0 +1,336 @@
+* Mutating Admission Webhooks can be used for defaulting, but:
+  * they are not adequate as it is not possible to set default values of newly added fields on GET because admission is not run in the storage layer when reading from etcd.
+  * webhooks are complex, both from the development/maintenance point of view and due to their non-trivial operational overhead. This makes them a no-go for many "not so ambitiously, professionally developed CRDs", e.g. in in-house enterprise environments.
+* _Structural schemas_ as defined in [KEP Vanilla OpenAPI Subset: Structural Schema](/keps/sig-api-machinery/20190425-structural-openapi.md) make defaulting a low hanging fruit.
+ 
+### Goals
+
+* add CustomResource defaulting at the correct position in the request pipeline 
+* allow definition of defaults via the OpenAPI v3 `default` field.
+
+### Non-Goals
+
+* allow non-constant defaults: native Golang code can of course set defaults which depends on other fields of a JSON object. This is out of scope here and would need some kind of defaulting webhook.
+* native-type declarative defaulting: this KEP is about CRDs. Though it might be desirable to support the same `// +default=<some-json-value>` tag and a mechanism in `k8s.io/apiserver` to evaluate defaults in native, generic registries, this is out-of-scope of the KEP though.
+
+## Proposal
+
+We assume the CRD has _structural schemas_ (as defined in [KEP Vanilla OpenAPI Subset: Structural Schema](/keps/sig-api-machinery/20190425-structural-openapi.md)).
+
+We propose to
+
+1. derive the value-validation-less variant of the structural schema (trivial by definition of _structural schema_) and 
+2. recursively follow the given CustomResource instance and the structural schema, applying defaults where an object field is 
+   * undefined (`_, ok := obj[field]; !ok`)
+   * `nil` if the field not nullable
+   * empty in case of lists and maps, and if nullable is not set.
+  
+   This means in detail for primitive types and no types (i.e. IntOrString or arbitrary JSON) in the structural schema:
+   
+   * `if v, ok := obj[fld]; !ok` => default
+   * `else if !nullable && v == nil` => default
+   
+   and for `array` type in the schema one of these:
+   
+   * `if v, ok := obj[fld]; !ok` => default
+   * `else if !nullable && v == nil` => default
+   * `else if array, ok := v.([]interface{}); !ok` => return deserialization error
+   * `else if len(array) == 0` => default
+  
+   and for `object` type in the schema:
+   
+   * `if v, ok := obj[fld]; !ok` => default
+   * `else if !nullable && v == nil` => default
+   * `else if object, ok := v.(map[string]interface{}); !ok` => return deserialization error
+   * `else if len(object) == 0` => default.
+   
+Hence, we never default zero numbers, integers or empty string if they are specified explicitly in the input.
```

Comment:
I'll just note that this is *slightly* different from core-type, where this happens if the field is omitempty.

Reply:
the more I think about it, the more I think this has to be wrong :-)

Reply:
go's zero values are a language oddity that bled into our API and combined with defaulting in unfortunate ways

fields in an unstructured object can distinguish between missing, null, and zero values. Treating them as the equivalent of go `*int32` and `*string` vars helps rationalize the behavior compared to in-tree types (and is the [guidance for optional fields](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required))

Reply:
> where this happens if the field is omitempty. (am I correct?)

omitempty doesn't control defaulting, it is typically zero-value checking (`0` for int, `""` for string, `nil` for *int and *string)

Reply:
how can we resolve this question?

To summarize:

1) native resources inherited the `null==undefined==empty` behaviour because a) Go (un)marshalling identifies null and undefined b) proto identifies null and empty.

2) in a fresh CRD world a) `null!=undefined` holds for the apiserver because in-memory JSON can distinguish b) `undefined!=empty` because we can enforce that via proper protobuf definitions in the future by one way or another (packaging, compare https://github.com/protocolbuffers/protobuf/issues/1606#issuecomment-281832148).

I tend to agree with @apelisse that we should go with the most simple semantics by default (which is also the minimal one). Anything else is surprising if you don't know the Golang marshalling/unmarshalling history.

Prospective additional tags `x-kubernetes-default-empty: true`, `x-kubernetes-default-null: true` would be an way to unblock us here. But would like to avoid them:

I have a hard time to imagine a good API which defaults empty lists/map. This feels counterintuitive and I as an API consumer am surprised when this happens. I consider handling like in [YamlTOJSON ](https://github.com/kubernetes-sigs/yaml/pull/14/files#diff-e51f89b39d976550831afb510a926be0R333) as a bug. Am not convinced we should let our API design be guided by library bugs.

The only reason for defaulting empty I can imagine here is the legacy one: CRDs which embed k/k types which have this odd semantics. `x-kubernetes-legacy-defaulting: true` or another painful name might be an idea.

Then there are nulls: a null which does not mean explicit null (and would be supposed therefore to be defaulted) most likely correspond to a missing `omitempty` or a missing pointer in Golang (`RawExtension`s are forced to write a null during marshalling. A `*RawExtension` could solve that).

tl/dr: clean semantics and the message: get your marshalling right and don't work around that with loose defaulting logic.

Reply:
To unblock this KEP: 

I suggest to use `default` as officially defined by OpenAPI v3 and suggested by @apelisse: only default on undefined.

We (can) add `x-kubernetes-legacy-default: <arbitrary-json>` as a second variant (warning: this is different than the bool in the previous comment. It is really a second defaulting field, mutually exclusive with the OpenAPI v3 one), in case we hit legacy situations where we want that loose matching behaviour.

Reply:
> I suggest to use `default` as officially defined by OpenAPI v3 and suggested by @apelisse: only default on undefined.

I agree